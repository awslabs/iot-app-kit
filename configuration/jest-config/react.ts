export default {
  setupFilesAfterEnv: ['jest-extended/all'],
  maxWorkers: '50%',
  moduleFileExtensions: ['js', 'ts', 'jsx', 'tsx'],
  collectCoverage: true,
  testEnvironment: 'jsdom',
  transform: {
    '.*\\.(tsx?|jsx?)$': [
      '@swc/jest',
      {
        jsc: {
          transform: {
            react: {
              runtime: 'automatic',
            },
          },
        },
      },
    ],
  },
  transformIgnorePatterns: [],
  moduleNameMapper: {
    '\\.(css|scss)$': 'identity-obj-proxy',
  },
};
