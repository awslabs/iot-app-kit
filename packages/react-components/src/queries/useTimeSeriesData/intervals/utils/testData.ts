import { sub, add, Duration } from 'date-fns';
import { Interval } from '../../types';

export const ANCHOR_DATE = new Date(1720488638812);

/**
 * All dates are anchored to the same point and can be converted via the slide duration
 */
export const fiveMinuteInterval: Interval = {
  end: ANCHOR_DATE,
  start: sub(ANCHOR_DATE, { minutes: 5 }),
};

export const tenMinuteInterval: Interval = {
  end: ANCHOR_DATE,
  start: sub(ANCHOR_DATE, { minutes: 10 }),
};

export const oneHourInterval: Interval = {
  end: ANCHOR_DATE,
  start: sub(ANCHOR_DATE, { hours: 1 }),
};

export const tenHourInterval: Interval = {
  end: ANCHOR_DATE,
  start: sub(ANCHOR_DATE, { hours: 10 }),
};

export const oneDayInterval: Interval = {
  end: ANCHOR_DATE,
  start: sub(ANCHOR_DATE, { days: 1 }),
};

export const tenDaysInterval: Interval = {
  end: ANCHOR_DATE,
  start: sub(ANCHOR_DATE, { days: 10 }),
};

export const slideInterval = (
  interval: Interval,
  duration: Duration
): Interval => {
  return {
    ...interval,
    end: add(interval.end, duration),
    start: add(interval.start, duration),
  };
};

export const shuffleIntervals = (intervals: Interval[]): Interval[] => {
  return intervals
    .map((value) => ({ value, sort: Math.random() }))
    .sort((a, b) => a.sort - b.sort)
    .map(({ value }) => value);
};
