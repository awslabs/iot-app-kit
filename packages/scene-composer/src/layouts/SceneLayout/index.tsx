export { default as ScenePanel } from './components/ScenePanel';
export { default as TabbedPanelContainer } from './components/TabbedPanelContainer';
export { default as FoldableContainer } from './components/FoldableContainer';
export { default as SceneLayout } from './SceneLayout';
export * from './components/utils';
