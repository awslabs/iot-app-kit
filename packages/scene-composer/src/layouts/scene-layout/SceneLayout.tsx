import React, { FC, Fragment, ReactNode, Suspense, useCallback, useContext } from 'react';
import { useIntl } from 'react-intl';
import styled, { ThemeContext } from 'styled-components';
import { Canvas, ThreeEvent } from '@react-three/fiber';
import { useContextBridge } from '@react-three/drei/core/useContextBridge';
import { isEmpty } from 'lodash';

import LoggingContext from '../../logger/react-logger/contexts/logging';
import MessageModal from '../../components/MessageModal';
import { MenuBar } from '../../components/MenuBar';
import { StaticLayout } from '../../components/StaticLayout';
import { WebGLCanvasManager } from '../../components/WebGLCanvasManager';
import { FloatingToolbar } from '../../components/toolbars';
import {
  SceneHierarchyPanel,
  SceneNodeInspectorPanel,
  SceneRulesPanel,
  SettingsPanel,
  TopBar,
} from '../../components/panels';
import { sceneComposerIdContext } from '../../sceneComposerIdContext';
import { useSceneDocument, useStore } from '../../store';
import { KnownComponentType, KnownSceneProperty } from '../../interfaces';
import { PoweredByMatterport } from '../../assets/svgs';

import LeftPanel from './components/LeftPanel';
import RightPanel from './components/RightPanel';

const UnselectableCanvas = styled(Canvas)`
  user-select: none;
  background: ${({ theme }) => {
    return theme.canvasBackground;
  }};
  z-index: 0;
`;

interface SceneLayoutProps {
  isViewing: boolean;
  onPointerMissed: (event: ThreeEvent<PointerEvent>) => void;
  LoadingView: ReactNode;
  showMessageModal: boolean;
}

const SceneLayout: FC<SceneLayoutProps> = ({ isViewing, onPointerMissed, LoadingView = null, showMessageModal }) => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  const valueDataBindingProvider = useStore(sceneComposerId)((state) => state.getEditorConfig)()
    .valueDataBindingProvider;
  const getComponentRefByType = useStore(sceneComposerId)((state) => state.getComponentRefByType);
  const componentNodeMap = useStore(sceneComposerId)((state) => state.document.componentNodeMap);
  const ContextBridge = useContextBridge(LoggingContext, sceneComposerIdContext, ThemeContext);
  const intl = useIntl();
  const { sceneLoaded, getSceneProperty } = useSceneDocument(sceneComposerId);
  const matterportModelId: string = getSceneProperty(KnownSceneProperty.MatterportModelId);

  const leftPanelEditModeProps = {
    [intl.formatMessage({ defaultMessage: 'Hierarchy', description: 'Panel Tab title' })]: <SceneHierarchyPanel />,
    [intl.formatMessage({ defaultMessage: 'Rules', description: 'Panel Tab title' })]: <SceneRulesPanel />,
    [intl.formatMessage({ defaultMessage: 'Settings', description: 'Panel Tab title' })]: (
      <SettingsPanel valueDataBindingProvider={valueDataBindingProvider} />
    ),
  };
  const leftPanelViewModeProps = {
    [intl.formatMessage({ defaultMessage: 'Hierarchy', description: 'Panel Tab title' })]: <SceneHierarchyPanel />,
  };
  const rightPanelProps = {
    [intl.formatMessage({ defaultMessage: 'Inspector', description: 'Panel Tab title' })]: <SceneNodeInspectorPanel />,
  };
  const leftPanel = <LeftPanel {...leftPanelEditModeProps} />;

  const viewingModeLeftPanel = <LeftPanel {...leftPanelViewModeProps} />;

  const rightPanel = <RightPanel {...rightPanelProps} />;

  const showTopBar = useCallback(() => {
    return !isEmpty(getComponentRefByType(KnownComponentType.MotionIndicator));
  }, [componentNodeMap]);

  return (
    <StaticLayout
      mainContent={
        <Fragment>
          <FloatingToolbar isViewing={isViewing} />
          {matterportModelId && <PoweredByMatterport matterportModelId={matterportModelId} />}
          <UnselectableCanvas shadows dpr={window.devicePixelRatio} onPointerMissed={onPointerMissed}>
            <ContextBridge>
              <Suspense fallback={LoadingView}>{!sceneLoaded ? null : <WebGLCanvasManager />}</Suspense>
            </ContextBridge>
          </UnselectableCanvas>
        </Fragment>
      }
      showModal={showMessageModal}
      modalContent={<MessageModal />}
      header={!isViewing && <MenuBar />}
      leftPanel={isViewing ? viewingModeLeftPanel : leftPanel}
      rightPanel={!isViewing && rightPanel}
      topBar={showTopBar() && <TopBar />}
    />
  );
};

export default SceneLayout;
