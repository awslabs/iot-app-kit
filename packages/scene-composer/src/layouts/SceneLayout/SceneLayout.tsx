import React, { FC, Fragment, ReactNode, Suspense, useContext, useMemo, useRef } from 'react';
import { useIntl } from 'react-intl';
import styled, { ThemeContext } from 'styled-components';
import { Canvas, ThreeEvent } from '@react-three/fiber';
import { useContextBridge } from '@react-three/drei/core/useContextBridge';
import { MatterportViewer, MpSdk } from '@matterport/r3f/dist';

import { setMatterportSdk } from '../../common/GlobalSettings';
import LoggingContext from '../../logger/react-logger/contexts/logging';
import MessageModal from '../../components/MessageModal';
import { MenuBar } from '../../components/MenuBar';
import { StaticLayout } from '../StaticLayout';
import { WebGLCanvasManager } from '../../components/WebGLCanvasManager';
import { FloatingToolbar } from '../../components/toolbars';
import {
  SceneHierarchyPanel,
  SceneNodeInspectorPanel,
  SceneRulesPanel,
  SettingsPanel,
  TopBar,
} from '../../components/panels';
import { sceneComposerIdContext } from '../../common/sceneComposerIdContext';
import { useSceneDocument, useStore } from '../../store';
import LogProvider from '../../logger/react-logger/log-provider';
import DefaultErrorFallback from '../../components/DefaultErrorFallback';
import { COMPOSER_FEATURES, ExternalLibraryConfig, KnownComponentType, MatterportConfig } from '../../interfaces';
import { CameraPreview } from '../../components/three-fiber/CameraPreview';
import useSelectedNode from '../../hooks/useSelectedNode';
import { findComponentByType } from '../../utils/nodeUtils';
import useFeature from '../../hooks/useFeature';

import LeftPanel from './components/LeftPanel';
import RightPanel from './components/RightPanel';
import CameraPreviewTrack from './components/CameraPreviewTrack';

const UnselectableCanvas = styled(Canvas)`
  user-select: none;
  background: ${({ theme }) => {
    return theme.canvasBackground;
  }};
  z-index: 0;
`;

const R3FWrapper = (props: {
  enableMatterport: boolean;
  matterportConfig?: MatterportConfig;
  children?: any;
  sceneLoaded?: boolean;
}) => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  const ContextBridge = useContextBridge(LoggingContext, sceneComposerIdContext, ThemeContext);

  if (!props.sceneLoaded) {
    return null;
  }

  if (props.enableMatterport && props.matterportConfig?.modelId) {
    return (
      <MatterportViewer
        assetBase={props.matterportConfig?.assetBase}
        m={props.matterportConfig?.modelId}
        applicationKey={props.matterportConfig?.applicationKey}
        connect-auth={props.matterportConfig?.accessToken}
        connect-provider='iot-twinmaker'
        onReady={(matterportSdk: MpSdk) => {
          // propagate this out elsewhere if you wish to useMatterportSdk() interface
          // to control this viewer from other non-3d ui
          // <MpSdkContext.Provider value={matterportSdk}>
          //  <Other2DComponents/>
          // <MpSdkContext.Provider/>
          console.log('MatterportViewer SDK ready!', matterportSdk);
          setMatterportSdk(sceneComposerId, matterportSdk);
        }}
        style={{ width: '100%', height: '100%' }}
        search={0}
        title={0}
        mt={0}
        play={1}
        mls={1}
      >
        <ContextBridge>
          <Suspense fallback={null}>{props.children}</Suspense>
        </ContextBridge>
      </MatterportViewer>
    );
  } else {
    return (
      <UnselectableCanvas shadows dpr={window.devicePixelRatio}>
        <ContextBridge>
          <Suspense fallback={null}>{props.children}</Suspense>
        </ContextBridge>
      </UnselectableCanvas>
    );
  }
};

interface SceneLayoutProps {
  isViewing: boolean;
  onPointerMissed: (event: ThreeEvent<PointerEvent>) => void;
  LoadingView: ReactNode;
  showMessageModal: boolean;
  externalLibraryConfig?: ExternalLibraryConfig;
}

const SceneLayout: FC<SceneLayoutProps> = ({
  isViewing,
  LoadingView = null,
  showMessageModal,
  externalLibraryConfig,
}) => {
  const sceneComposerId = useContext(sceneComposerIdContext);

  const valueDataBindingProvider = useStore(sceneComposerId)((state) => state.getEditorConfig)()
    .valueDataBindingProvider;
  const ContextBridge = useContextBridge(LoggingContext, sceneComposerIdContext, ThemeContext);
  const intl = useIntl();
  const { sceneLoaded } = useSceneDocument(sceneComposerId);

  const renderDisplayRef = useRef<HTMLDivElement>(null!);

  const selectedNode = useSelectedNode();

  const shouldShowPreview = useMemo(() => {
    return isViewing ? false : !!findComponentByType(selectedNode.selectedSceneNode, KnownComponentType.Camera);
  }, [selectedNode]);

  const [{ variation: matterportFeature }] = useFeature(COMPOSER_FEATURES.Matterport);

  const leftPanelEditModeProps = {
    [intl.formatMessage({ defaultMessage: 'Hierarchy', description: 'Panel Tab title' })]: <SceneHierarchyPanel />,
    [intl.formatMessage({ defaultMessage: 'Rules', description: 'Panel Tab title' })]: <SceneRulesPanel />,
    [intl.formatMessage({ defaultMessage: 'Settings', description: 'Panel Tab title' })]: (
      <SettingsPanel valueDataBindingProvider={valueDataBindingProvider} />
    ),
  };
  const leftPanelViewModeProps = {
    [intl.formatMessage({ defaultMessage: 'Hierarchy', description: 'Panel Tab title' })]: <SceneHierarchyPanel />,
    [intl.formatMessage({ defaultMessage: 'Settings', description: 'Panel Tab title' })]: (
      <SettingsPanel valueDataBindingProvider={valueDataBindingProvider} />
    ),
  };
  const rightPanelProps = {
    [intl.formatMessage({ defaultMessage: 'Inspector', description: 'Panel Tab title' })]: <SceneNodeInspectorPanel />,
  };

  const leftPanel = <LeftPanel {...leftPanelEditModeProps} />;

  const viewingModeLeftPanel = <LeftPanel {...leftPanelViewModeProps} />;

  const rightPanel = <RightPanel {...rightPanelProps} />;

  return (
    <StaticLayout
      mainContent={
        <Fragment>
          <LogProvider namespace='SceneLayout' ErrorView={DefaultErrorFallback}>
            <FloatingToolbar isViewing={isViewing} />
            {/* {matterportModelId && <PoweredByMatterport matterportModelId={matterportModelId} />} */}
            {/*
            // TODO(mp): three upgrade type mismatch much unreadable, triage further.
            <UnselectableCanvas shadows dpr={window.devicePixelRatio} onPointerMissed={onPointerMissed}>
            */}
            <ContextBridge>
              <Suspense fallback={LoadingView}>
                {shouldShowPreview && (
                  <CameraPreviewTrack ref={renderDisplayRef} title={selectedNode.selectedSceneNode?.name} />
                )}
                <R3FWrapper
                  enableMatterport={matterportFeature === 'T1' && !!externalLibraryConfig?.matterport?.modelId}
                  sceneLoaded={sceneLoaded}
                  matterportConfig={externalLibraryConfig?.matterport}
                >
                  {/* TODO: Add loading view */}
                  {!sceneLoaded ? null : (
                    <Fragment>
                      <WebGLCanvasManager />
                      {shouldShowPreview && <CameraPreview track={renderDisplayRef} />}
                    </Fragment>
                  )}
                </R3FWrapper>
              </Suspense>
            </ContextBridge>
          </LogProvider>
        </Fragment>
      }
      showModal={showMessageModal}
      modalContent={<MessageModal />}
      header={!isViewing && <MenuBar />}
      leftPanel={isViewing ? viewingModeLeftPanel : leftPanel}
      rightPanel={!isViewing && rightPanel}
      topBar={<TopBar />}
    />
  );
};

export default SceneLayout;
