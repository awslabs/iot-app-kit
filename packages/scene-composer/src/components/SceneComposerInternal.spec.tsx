import * as React from 'react';
import renderer, { act } from 'react-test-renderer';
import str2ab from 'string-to-arraybuffer';
import flushPromises from 'flush-promises';

import { SceneComposerInternal, SceneComposerApi, useSceneComposerApi } from '..';
import * as SceneLayoutComponents from '../layouts/SceneLayout';
import { invalidTestScenes, testScenes } from '../../tests/testData';

jest.mock('../layouts/StaticLayout', () => ({
  StaticLayout: 'StaticLayout',
}));

jest.mock((window as any).ResizeObserver, () => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn(),
}));

// @ts-ignore
jest.mock('scheduler', () => require('scheduler/unstable_mock'));
jest.mock('resize-observer-polyfill', () => {
  return ResizeObserver;
});

function createSceneLoaderMock(sceneContent: string) {
  return {
    getSceneUri: () => Promise.resolve('file://test.json'),
    getSceneUrl: () => Promise.resolve('file://test.json'),
    getSceneObject: (uri: string) => Promise.resolve(str2ab(sceneContent)),
  };
}

describe('SceneComposerInternal', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly with an empty scene in editing mode', async () => {
    let container;
    await act(async () => {
      container = renderer.create(
        <SceneComposerInternal config={{ mode: 'Editing' }} sceneLoader={createSceneLoaderMock('')} />,
      );

      await flushPromises();
    });

    // shows the scene hierarchy browser
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with an empty scene in viewing mode', async () => {
    let container;
    await act(async () => {
      container = renderer.create(
        <SceneComposerInternal config={{ mode: 'Viewing' }} sceneLoader={createSceneLoaderMock('')} />,
      );

      await flushPromises();
    });

    // shows the scene hierarchy browser
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with a valid scene in editing mode', async () => {
    let container;
    await act(async () => {
      container = renderer.create(
        <SceneComposerInternal config={{ mode: 'Editing' }} sceneLoader={createSceneLoaderMock(testScenes.scene1)} />,
      );

      await flushPromises();
    });

    // shows the scene hierarchy browser
    expect(container).toMatchSnapshot();
  });

  it('should render warning when minor version is newer', async () => {
    let container;
    await act(async () => {
      container = renderer.create(
        <SceneComposerInternal
          config={{ mode: 'Editing' }}
          sceneLoader={createSceneLoaderMock(invalidTestScenes.unsupportedMinorVersionScene)}
        />,
      );

      await flushPromises();
    });

    expect(container).toMatchSnapshot();
  });

  it('should render error when major version is newer', async () => {
    let container;
    await act(async () => {
      container = renderer.create(
        <SceneComposerInternal
          config={{ mode: 'Editing' }}
          sceneLoader={createSceneLoaderMock(invalidTestScenes.unsupportedMajorVersion)}
        />,
      );

      await flushPromises();
    });

    expect(container).toMatchSnapshot();
  });

  it('should render error when specVersion is invalid', async () => {
    let container;
    await act(async () => {
      container = renderer.create(
        <SceneComposerInternal
          config={{ mode: 'Editing' }}
          sceneLoader={createSceneLoaderMock(invalidTestScenes.invalidSpecVersionScene)}
        />,
      );

      await flushPromises();
    });

    expect(container).toMatchSnapshot();
  });

  it('should support rendering multiple valid scenes', async () => {
    let container;
    await act(async () => {
      container = renderer.create(
        <div>
          <SceneComposerInternal config={{ mode: 'Editing' }} sceneLoader={createSceneLoaderMock(testScenes.scene1)} />
          <SceneComposerInternal config={{ mode: 'Editing' }} sceneLoader={createSceneLoaderMock(testScenes.scene2)} />
        </div>,
      );

      await flushPromises();
    });

    // verify that 2 different scenes are rendered
    expect(container).toMatchSnapshot();
  });

  it('should render both valid and invalid scene correctly', async () => {
    let container;
    await act(async () => {
      container = renderer.create(
        <div>
          <SceneComposerInternal
            config={{ mode: 'Editing' }}
            sceneLoader={createSceneLoaderMock(invalidTestScenes.invalidJson)}
          />
          <SceneComposerInternal config={{ mode: 'Editing' }} sceneLoader={createSceneLoaderMock(testScenes.scene1)} />
        </div>,
      );

      await flushPromises();
    });

    expect(container).toMatchSnapshot();
  });

  it('should render a default error view when loading a bad scene content', async () => {
    let container;
    await act(async () => {
      container = renderer.create(
        <SceneComposerInternal
          config={{ mode: 'Editing' }}
          sceneLoader={createSceneLoaderMock(invalidTestScenes.invalidJson)}
        />,
      );

      await flushPromises();
    });

    expect(container).toMatchSnapshot();
  });

  it('should render a default error view when unknown error happens', async () => {
    const mocked = jest.spyOn(SceneLayoutComponents, 'SceneLayout').mockImplementation(() => {
      throw new Error('failed to render');
    });

    const container = renderer.create(
      <SceneComposerInternal config={{ mode: 'Editing' }} sceneLoader={createSceneLoaderMock('')} />,
    );

    await flushPromises();

    expect(container).toMatchSnapshot();

    mocked.mockRestore();
  });

  describe('useSceneComposerApi', () => {
    it('should return an api object', async () => {
      let sut: SceneComposerApi | null = null;

      const TestComponent = () => {
        const sceneComposerId = 'test';
        sut = useSceneComposerApi('test');

        return (
          <SceneComposerInternal
            sceneComposerId={sceneComposerId}
            config={{ mode: 'Editing' }}
            sceneLoader={createSceneLoaderMock(testScenes.scene1)}
          />
        );
      };

      await act(async () => {
        renderer.create(<TestComponent />);

        await flushPromises();
      });

      expect(sut).toHaveProperty('setCameraTarget');
    });
  });
});
