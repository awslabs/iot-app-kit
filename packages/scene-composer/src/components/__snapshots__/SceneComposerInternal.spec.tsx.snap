// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SceneComposerInternal should render a default error view when loading a bad scene content 1`] = `
<StaticLayout
  header={<MenuBar />}
  leftPanel={
    <LeftPanel
      Hierarchy={<SceneHierarchyPanel />}
      Rules={<SceneRulesPanel />}
      Settings={<SettingsPanel />}
    />
  }
  mainContent={
    <React.Fragment>
      <LogProvider
        ErrorView={[Function]}
        namespace="SceneLayout"
      >
        <FloatingToolbar
          isViewing={false}
        />
        <Unknown>
          <React.Suspense
            fallback={
              <Provider>
                <LoadingProgress />
              </Provider>
            }
          >
            <R3FWrapper
              enableMatterport={false}
              sceneLoaded={true}
            >
              <React.Fragment>
                <WebGLCanvasManager />
              </React.Fragment>
            </R3FWrapper>
          </React.Suspense>
        </Unknown>
      </LogProvider>
    </React.Fragment>
  }
  modalContent={<MessageModal />}
  rightPanel={
    <RightPanel
      Inspector={<SceneNodeInspectorPanel />}
    />
  }
  showModal={true}
  topBar={<TopBar />}
/>
`;

exports[`SceneComposerInternal should render a default error view when unknown error happens 1`] = `
<StaticLayout
  modalContent={
    <Styled(MockContainer)
      header={
        <MockHeader
          variant="h2"
        >
          Error
        </MockHeader>
      }
    >
      <MockTextContent>
        <p>
          failed to render
        </p>
      </MockTextContent>
    </Styled(MockContainer)>
  }
  showModal={true}
/>
`;

exports[`SceneComposerInternal should render both valid and invalid scene correctly 1`] = `
<div>
  <StaticLayout
    header={<MenuBar />}
    leftPanel={
      <LeftPanel
        Hierarchy={<SceneHierarchyPanel />}
        Rules={<SceneRulesPanel />}
        Settings={<SettingsPanel />}
      />
    }
    mainContent={
      <React.Fragment>
        <LogProvider
          ErrorView={[Function]}
          namespace="SceneLayout"
        >
          <FloatingToolbar
            isViewing={false}
          />
          <Unknown>
            <React.Suspense
              fallback={
                <Provider>
                  <LoadingProgress />
                </Provider>
              }
            >
              <R3FWrapper
                enableMatterport={false}
                sceneLoaded={true}
              >
                <React.Fragment>
                  <WebGLCanvasManager />
                </React.Fragment>
              </R3FWrapper>
            </React.Suspense>
          </Unknown>
        </LogProvider>
      </React.Fragment>
    }
    modalContent={<MessageModal />}
    rightPanel={
      <RightPanel
        Inspector={<SceneNodeInspectorPanel />}
      />
    }
    showModal={true}
    topBar={<TopBar />}
  />
  <StaticLayout
    header={<MenuBar />}
    leftPanel={
      <LeftPanel
        Hierarchy={<SceneHierarchyPanel />}
        Rules={<SceneRulesPanel />}
        Settings={<SettingsPanel />}
      />
    }
    mainContent={
      <React.Fragment>
        <LogProvider
          ErrorView={[Function]}
          namespace="SceneLayout"
        >
          <FloatingToolbar
            isViewing={false}
          />
          <Unknown>
            <React.Suspense
              fallback={
                <Provider>
                  <LoadingProgress />
                </Provider>
              }
            >
              <R3FWrapper
                enableMatterport={false}
                sceneLoaded={true}
              >
                <React.Fragment>
                  <WebGLCanvasManager />
                </React.Fragment>
              </R3FWrapper>
            </React.Suspense>
          </Unknown>
        </LogProvider>
      </React.Fragment>
    }
    modalContent={<MessageModal />}
    rightPanel={
      <RightPanel
        Inspector={<SceneNodeInspectorPanel />}
      />
    }
    showModal={false}
    topBar={<TopBar />}
  />
</div>
`;

exports[`SceneComposerInternal should render correctly with a valid scene in editing mode 1`] = `
<StaticLayout
  header={<MenuBar />}
  leftPanel={
    <LeftPanel
      Hierarchy={<SceneHierarchyPanel />}
      Rules={<SceneRulesPanel />}
      Settings={<SettingsPanel />}
    />
  }
  mainContent={
    <React.Fragment>
      <LogProvider
        ErrorView={[Function]}
        namespace="SceneLayout"
      >
        <FloatingToolbar
          isViewing={false}
        />
        <Unknown>
          <React.Suspense
            fallback={
              <Provider>
                <LoadingProgress />
              </Provider>
            }
          >
            <R3FWrapper
              enableMatterport={false}
              sceneLoaded={true}
            >
              <React.Fragment>
                <WebGLCanvasManager />
              </React.Fragment>
            </R3FWrapper>
          </React.Suspense>
        </Unknown>
      </LogProvider>
    </React.Fragment>
  }
  modalContent={<MessageModal />}
  rightPanel={
    <RightPanel
      Inspector={<SceneNodeInspectorPanel />}
    />
  }
  showModal={false}
  topBar={<TopBar />}
/>
`;

exports[`SceneComposerInternal should render correctly with an empty scene in editing mode 1`] = `
<StaticLayout
  header={<MenuBar />}
  leftPanel={
    <LeftPanel
      Hierarchy={<SceneHierarchyPanel />}
      Rules={<SceneRulesPanel />}
      Settings={<SettingsPanel />}
    />
  }
  mainContent={
    <React.Fragment>
      <LogProvider
        ErrorView={[Function]}
        namespace="SceneLayout"
      >
        <FloatingToolbar
          isViewing={false}
        />
        <Unknown>
          <React.Suspense
            fallback={
              <Provider>
                <LoadingProgress />
              </Provider>
            }
          >
            <R3FWrapper
              enableMatterport={false}
            />
          </React.Suspense>
        </Unknown>
      </LogProvider>
    </React.Fragment>
  }
  modalContent={<MessageModal />}
  rightPanel={
    <RightPanel
      Inspector={<SceneNodeInspectorPanel />}
    />
  }
  showModal={false}
  topBar={<TopBar />}
/>
`;

exports[`SceneComposerInternal should render correctly with an empty scene in viewing mode 1`] = `
<StaticLayout
  header={false}
  leftPanel={
    <LeftPanel
      Hierarchy={<SceneHierarchyPanel />}
      Settings={<SettingsPanel />}
    />
  }
  mainContent={
    <React.Fragment>
      <LogProvider
        ErrorView={[Function]}
        namespace="SceneLayout"
      >
        <FloatingToolbar
          isViewing={true}
        />
        <Unknown>
          <React.Suspense
            fallback={
              <Provider>
                <LoadingProgress />
              </Provider>
            }
          >
            <R3FWrapper
              enableMatterport={false}
            />
          </React.Suspense>
        </Unknown>
      </LogProvider>
    </React.Fragment>
  }
  modalContent={<MessageModal />}
  rightPanel={false}
  showModal={false}
  topBar={<TopBar />}
/>
`;

exports[`SceneComposerInternal should render error when major version is newer 1`] = `
<StaticLayout
  header={<MenuBar />}
  leftPanel={
    <LeftPanel
      Hierarchy={<SceneHierarchyPanel />}
      Rules={<SceneRulesPanel />}
      Settings={<SettingsPanel />}
    />
  }
  mainContent={
    <React.Fragment>
      <LogProvider
        ErrorView={[Function]}
        namespace="SceneLayout"
      >
        <FloatingToolbar
          isViewing={false}
        />
        <Unknown>
          <React.Suspense
            fallback={
              <Provider>
                <LoadingProgress />
              </Provider>
            }
          >
            <R3FWrapper
              enableMatterport={false}
              sceneLoaded={true}
            >
              <React.Fragment>
                <WebGLCanvasManager />
              </React.Fragment>
            </R3FWrapper>
          </React.Suspense>
        </Unknown>
      </LogProvider>
    </React.Fragment>
  }
  modalContent={<MessageModal />}
  rightPanel={
    <RightPanel
      Inspector={<SceneNodeInspectorPanel />}
    />
  }
  showModal={true}
  topBar={<TopBar />}
/>
`;

exports[`SceneComposerInternal should render error when specVersion is invalid 1`] = `
<StaticLayout
  header={<MenuBar />}
  leftPanel={
    <LeftPanel
      Hierarchy={<SceneHierarchyPanel />}
      Rules={<SceneRulesPanel />}
      Settings={<SettingsPanel />}
    />
  }
  mainContent={
    <React.Fragment>
      <LogProvider
        ErrorView={[Function]}
        namespace="SceneLayout"
      >
        <FloatingToolbar
          isViewing={false}
        />
        <Unknown>
          <React.Suspense
            fallback={
              <Provider>
                <LoadingProgress />
              </Provider>
            }
          >
            <R3FWrapper
              enableMatterport={false}
              sceneLoaded={true}
            >
              <React.Fragment>
                <WebGLCanvasManager />
              </React.Fragment>
            </R3FWrapper>
          </React.Suspense>
        </Unknown>
      </LogProvider>
    </React.Fragment>
  }
  modalContent={<MessageModal />}
  rightPanel={
    <RightPanel
      Inspector={<SceneNodeInspectorPanel />}
    />
  }
  showModal={true}
  topBar={<TopBar />}
/>
`;

exports[`SceneComposerInternal should render warning when minor version is newer 1`] = `
<StaticLayout
  header={<MenuBar />}
  leftPanel={
    <LeftPanel
      Hierarchy={<SceneHierarchyPanel />}
      Rules={<SceneRulesPanel />}
      Settings={<SettingsPanel />}
    />
  }
  mainContent={
    <React.Fragment>
      <LogProvider
        ErrorView={[Function]}
        namespace="SceneLayout"
      >
        <FloatingToolbar
          isViewing={false}
        />
        <Unknown>
          <React.Suspense
            fallback={
              <Provider>
                <LoadingProgress />
              </Provider>
            }
          >
            <R3FWrapper
              enableMatterport={false}
              sceneLoaded={true}
            >
              <React.Fragment>
                <WebGLCanvasManager />
              </React.Fragment>
            </R3FWrapper>
          </React.Suspense>
        </Unknown>
      </LogProvider>
    </React.Fragment>
  }
  modalContent={<MessageModal />}
  rightPanel={
    <RightPanel
      Inspector={<SceneNodeInspectorPanel />}
    />
  }
  showModal={true}
  topBar={<TopBar />}
/>
`;

exports[`SceneComposerInternal should support rendering multiple valid scenes 1`] = `
<div>
  <StaticLayout
    header={<MenuBar />}
    leftPanel={
      <LeftPanel
        Hierarchy={<SceneHierarchyPanel />}
        Rules={<SceneRulesPanel />}
        Settings={<SettingsPanel />}
      />
    }
    mainContent={
      <React.Fragment>
        <LogProvider
          ErrorView={[Function]}
          namespace="SceneLayout"
        >
          <FloatingToolbar
            isViewing={false}
          />
          <Unknown>
            <React.Suspense
              fallback={
                <Provider>
                  <LoadingProgress />
                </Provider>
              }
            >
              <R3FWrapper
                enableMatterport={false}
                sceneLoaded={true}
              >
                <React.Fragment>
                  <WebGLCanvasManager />
                </React.Fragment>
              </R3FWrapper>
            </React.Suspense>
          </Unknown>
        </LogProvider>
      </React.Fragment>
    }
    modalContent={<MessageModal />}
    rightPanel={
      <RightPanel
        Inspector={<SceneNodeInspectorPanel />}
      />
    }
    showModal={false}
    topBar={<TopBar />}
  />
  <StaticLayout
    header={<MenuBar />}
    leftPanel={
      <LeftPanel
        Hierarchy={<SceneHierarchyPanel />}
        Rules={<SceneRulesPanel />}
        Settings={<SettingsPanel />}
      />
    }
    mainContent={
      <React.Fragment>
        <LogProvider
          ErrorView={[Function]}
          namespace="SceneLayout"
        >
          <FloatingToolbar
            isViewing={false}
          />
          <Unknown>
            <React.Suspense
              fallback={
                <Provider>
                  <LoadingProgress />
                </Provider>
              }
            >
              <R3FWrapper
                enableMatterport={false}
                sceneLoaded={true}
              >
                <React.Fragment>
                  <WebGLCanvasManager />
                </React.Fragment>
              </R3FWrapper>
            </React.Suspense>
          </Unknown>
        </LogProvider>
      </React.Fragment>
    }
    modalContent={<MessageModal />}
    rightPanel={
      <RightPanel
        Inspector={<SceneNodeInspectorPanel />}
      />
    }
    showModal={false}
    topBar={<TopBar />}
  />
</div>
`;
